lib/src/main/java/com/cloudbees/groovy/cps/impl/CpsDelegatingClosure.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.groovy.cps.impl;
+
+import com.cloudbees.groovy.cps.Continuable;
+import com.cloudbees.groovy.cps.Outcome;
+import groovy.lang.Closure;
+import groovy.lang.MetaClassImpl;
+import org.codehaus.groovy.classgen.asm.ClosureWriter;
+import org.codehaus.groovy.runtime.CurriedClosure;
+import org.codehaus.groovy.runtime.typehandling.DefaultTypeTransformation;
+
+import javax.annotation.Nonnull;
+
+public class CpsDelegatingClosure extends Closure {
+    private static final Object[] EMPTY_OBJECT_ARRAY = {};
+
+    private final CpsClosure cpsClosure;
+
+    public CpsDelegatingClosure(@Nonnull CpsClosure cpsClosure) {
+        super(cpsClosure.getOwner(), cpsClosure.getThisObject());
+        this.cpsClosure = cpsClosure;
+    }
+
+    @Override
+    public Object call() {
+        return call(EMPTY_OBJECT_ARRAY);
+    }
+
+    @Override
+    public Object call(Object... args) {
+        try {
+            return cpsClosure.call(args);
+        } catch (CpsCallableInvocation inv) {
+            Continuable continuable = new Continuable(inv.asBlock());
+            Outcome o = continuable.run0(new Outcome(null, null), Continuable.categories);
+            if (o.isSuccess()) {
+                return o.getNormal();
+            } else {
+                throw new RuntimeException(o.getAbnormal());
+            }
```

Comment:
I spent a bit of time looking into this and I think the reason the patch works here but not in `workflow-cps` is related to the safepoint mechanism. As it stands, this PR only executes the CPS-transformed closure up to the first safepoint. With a hack like the following, basic tests against `workflow-cps` start to pass:

```suggestion
            Outcome o = new Outcome(null, null);
            while (true) {
                o = continuable.run0(o, Continuable.categories);
                if (o.isSuccess()) {
                    Object normal = o.getNormal();
                    if (normal != null && normal.getClass().getName().equals("org.jenkinsci.plugins.workflow.cps.Safepoint")) {
                        o = new Outcome(null, null);
                    } else {
                        return normal;
                    }
                } else {
                    throw new RuntimeException(o.getAbnormal());
                }
            }
```

That said, I'm not sure how we could get something like `[-2, 1, -4, 3].sort({i -> sleep 1; i * i})` to execute in a way that makes sense using this approach. Maybe that's not a big deal though, since most closures used for these methods probably don't call Pipeline steps at all, let alone async Pipeline steps.

If I have some more time I will look into translating the methods that have issues into a variant that just throws an exception if they are called with a `CpsClosure` to at least make the current issues easier for users to understand.

Reply:
> throws an exception if they are called with a `CpsClosure`

You do not think https://github.com/jenkinsci/workflow-cps-plugin/pull/211 would suffice?

Reply:
That system works fine, but I think it would still be nice to have a more explicit fail-fast warning for cases where we know the method is not supported, e.g. `UnsupportedOperationException: DefaultGroovyMethods.sort(Foo, Closure) is not supported by Pipeline, see https://www.jenkins.io/doc/book/pipeline/cps-method-mismatches/ for details.`